psamm/moma.py

```diff
@@ -79,84 +100,332 @@ def __init__(self, model, solver):
+        if not result:
+            raise MOMAError('Unable to solve LP2 MOMA: {}'.format(
+                result.status))
+
+        # Go back and manually delete all the constaints
+        constr.delete()
+
+    def minimize_l1(self, objective, wt_obj, wt_fluxes):
+        """Implemetnation of the lin_MOMA2 algorithm.
+
+        We try to maximize biomass, by keeping the fluxes relitively the same
+        as the wildtype. This helps avoid the assumption that an organism will
+        perform optimally directly after removing a gene.
+
+        Creates the constraint that the we select the optimal flux vector that
+        is closest to the wildtype.
 
-            self.constraints(zs >= vs_wt - vs, vs_wt - vs >= -zs)
+        Minimizes sum of |wild type - knockout|
+
+        Args:
+            objective: Biomass reaction for the model.
+            wt_obj: The wild type biomass value found using FBA.
+            wt_fluxes: All the wt_fluxes found by FBA in the form of a
+                dictionary. Use :meth: _get_fba_flux(objective) to return
+                a dictionary.
+        """
+        # These are all of our non eachange reactions
+        reactions = set(self._adjustment_reactions())
+        # First times in the loop
+        if self._z is None:
+            self._z = z = self._prob.namespace()
+            z.define(reactions, lower=0)
+        # After the first time, we can just use the same z object as before
         else:
             z = self._z
 
+        # Get the v (knockout) and v_wt (wild type) model and all the
+        # reactions 'variables' associated with it
+        v = self._v
+        v_wt = self._v_wt
+
+        # Create a constraints object for this model
+        constr = self.constraints()
+
+        # Loop through all the flux reactions
+        for f_reaction, f_value in iteritems(wt_fluxes):
+            # Makes sure the reaction we are trying to put a constraint on
+            # is a non exchange reaction
+            if f_reaction in reactions:
+                # Add the constraint that finds the optimal solution, such
+                # that the difference between the wildtype flux
+                # is similar to the knockout flux.
+                constr.add(
+                    z(f_reaction) >= v_wt(f_reaction) - v(f_reaction),
+                    v_wt(f_reaction) - v(f_reaction) >= -z(f_reaction))
+
+        # Set the objective for the sum of z
         self._prob.set_objective(z.sum(self._adjustment_reactions()))
 
+        # Set a constraint that says the biomass flux of the wild type is
+        # the same as the FBA analysis
         v_wt_obj = self._v_wt(objective)
         with self.constraints(v_wt_obj == wt_obj):
             result = self._solve(lp.ObjectiveSense.Minimize)
 
+        # Raise any errors
         if not result:
             raise MOMAError('Unable to solve L1 MOMA: {}'.format(
                 result.status))
```

Comment:
Here, the constraints do not seem to be removed after solving. Should be handled with `try`-block as above.